src/Common/src/System/Text/StringOrCharArray.cs

```diff
@@ -0,0 +1,178 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+namespace System.Text
+{
+    /// <summary>
+    /// Discrimated union of a string and a char array/offset/count.  Enables looking up
+    /// a portion of a char array as a key in a dictionary of string keys without having to
+    /// allocate/copy the chars into a new string.  This comes at the expense of an extra
+    /// reference field + two Int32s per key in the size of the dictionary's Entry array.
+    /// </summary>
+    internal struct StringOrCharArray : IEquatable<StringOrCharArray>
+    {
+        public readonly string String;
+
+        public readonly char[] CharArray;
+        public readonly int CharArrayOffset;
+        public readonly int CharArrayCount;
+
+        public StringOrCharArray(string s)
+        {
+            String = s;
+
+            CharArray = null;
+            CharArrayOffset = 0;
+            CharArrayCount = 0;
+
+            DebugValidate();
+        }
+
+        public StringOrCharArray(char[] charArray, int charArrayIndex, int charArrayOffset)
+        {
+            String = null;
+
+            CharArray = charArray;
+            CharArrayOffset = charArrayIndex;
+            CharArrayCount = charArrayOffset;
+
+            DebugValidate();
+        }
+
+        public static implicit operator StringOrCharArray(string value)
+        {
+            return new StringOrCharArray(value);
+        }
+
+        public int Length
+        {
+            get
+            {
+                DebugValidate();
+                return String != null ? String.Length : CharArrayCount;
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            return
+                obj is StringOrCharArray &&
+                Equals((StringOrCharArray)obj);
+        }
+
+        public unsafe bool Equals(StringOrCharArray other)
+        {
+            this.DebugValidate();
+            other.DebugValidate();
+
+            if (this.String != null)
+            {
+                // String vs String
+                if (other.String != null)
+                {
+                    return StringComparer.Ordinal.Equals(this.String, other.String);
+                }
+
+                // String vs CharArray
+                if (this.String.Length != other.CharArrayCount)
+                    return false;
+
+                for (int i = 0; i < this.String.Length; i++)
+                {
+                    if (this.String[i] != other.CharArray[other.CharArrayOffset + i])
+                        return false;
+                }
+
+                return true;
+            }
+
+            // CharArray vs CharArray
+            if (other.CharArray != null)
+            {
+                if (this.CharArrayCount != other.CharArrayCount)
+                    return false;
+
+                for (int i = 0; i < this.CharArrayCount; i++)
```

Comment:
With current usage it'll never be empty.  As with the other, if usage changes such that such cases become more common and performance of those cases becomes a bottle-neck, they could certainly be optimized.